### PR TITLE
CPLAT-9327 Improve forwardRef documentation, add displayName argument

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
> __DEPENDS ON [UNRELEASED CHANGES IN REACT-DART](https://github.com/cleandart/react-dart/pull/251)__

## Motivation
While grokking our `forwardRef` API, I found myself wanting some implementation examples which could not be found in the documentation comment.

I also noticed that consumers had no way to define a `displayName` for the wrapper HOC created by `forwardRef`.

## Changes
These changes address both issues by adding extensive examples in the documentation comment for `forwardRef` that mirror the analogous [ReactJS forwardRef documentation](https://reactjs.org/docs/forwarding-refs.html), and by adding a `displayName` argument to the function which then gets set on the JS side via a call to `defineProperty`.

#### Release Notes
Allow custom displayName values for `forwardRef` wrapper components.

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: @greglittlefield-wf @joebingham-wk @sydneyjodon-wk

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Client Platform member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/teams/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: web/component2/src/demos/forward_ref.dart [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  @override
+  render() {
+    return (props.builder()
+      ..modifyProps(addUnconsumedDomProps)
+      ..ref = props._forwardedRef
+    )(props.children);
+  }
+}
+
+//class _LogProps extends react.Component2 {
```
Comment (greglittlefield-wf) [inline web/component2/src/demos/forward_ref.dart]: Are these comments supposed to be here?

**Comment by greglittlefield-wf** [comment]
Also, should this go into a function components WIP branch if it's dependent on changes in the react-dart 5.4.0-wip?

File: web/component2/src/demos/forward_ref.dart (hunk 1 above)
Comment (smaifullerton-wk) [reply web/component2/src/demos/forward_ref.dart]: Nope per Aaron, removed.

File: lib/src/component/ref_util.dart [hunk 2]
```diff
@@ -52,52 +54,153 @@ Ref<T> createRef<T>() {
...
+/// final LogProps = forwardRef<LogPropsProps>((props, ref) {
+///   return (_LogProps()
+///     ..addProps(props)
+///     .._forwardedRef = ref
+///   )('Click me!');
+/// })(_LogProps);
+///
+/// UiFactory<LogPropsProps> _LogProps = _$_LogProps;
+///
+/// mixin LogPropsProps on UiProps {
```
Comment (smaifullerton-wk) [inline lib/src/component/ref_util.dart:127]: Please double check the boilerplate here - I updated this example from what it was previously.

**Comment by aaronlademann-wf** [comment]
@smaifullerton-wk I don't think the stuff you need from the `react` package is in the 5.4.0 release.  You'll still need to have that `5.4.0/improve-forwardref-examples-and-typing` dep override until its released.

**Comment by aaronlademann-wf** [comment]
> Also, should this go into a function components WIP branch if it's dependent on changes in the react-dart 5.4.0-wip?

@greglittlefield-wf is there a function components wip branch here? I see both this - and Sydney's function component PRs with a base branch of `master`...

**Comment by greglittlefield-wf** [comment]
@aaronlademann-wf I don't believe there is one, no, but we could make one!

**Comment by aaronlademann-wf** [comment]
@greglittlefield-wf created the branch / updated the base branch of this PR. (I'll update Sydney's as well)

**Pull request merged**